plutus-core/test/Evaluation/Golden.hs

```diff
@@ -151,104 +152,163 @@ lteExpr = mkIterApp () lte [eleven, twentytwo]
+iteAtIntegerArrowIntegerAppliedApplied :: Term TyName Name DefaultUni DefaultFun ()
+iteAtIntegerArrowIntegerAppliedApplied =  Apply () iteAtIntegerArrowIntegerApplied1 twentytwo
+
+-- { (builtin ifThenElse) (lam a . a -> a) }
+-- Evaluation should succeed, but typechecking should fail with a kind error.
+-- The built-in function machinery does allow builtins which are polymorphic
+-- over higher-kinded types, but `ifThenElse` can only be instantiated with
+-- types of kind *.
 iteAtHigherKind :: Term TyName Name DefaultUni DefaultFun ()
-iteAtHigherKind = TyInst () ite (TyForall () a (Type ()) aArrowA)
+iteAtHigherKind = iteAt (TyLam () a (Type ()) aArrowA)
     where a = TyName (Name "a" (Unique 0))
           aArrowA = TyFun () (TyVar () a) (TyVar () a)
 
--- [ (builtin ifThenElse) (11<=22) ]: should fail because ifThenElse isn't instantiated
--- Type expected, term supplied.
-iteUninstantiatedOneArg :: Term TyName Name DefaultUni DefaultFun ()
-iteUninstantiatedOneArg  = Apply () ite lteExpr
-
--- [ { (builtin ifThenElse) (con integer) } (11<=22)]: should succeed.
-iteAtIntegerOneArg :: Term TyName Name DefaultUni DefaultFun ()
-iteAtIntegerOneArg = Apply () iteAtInteger lteExpr
-
--- [ { (builtin ifThenElse) (con string) } (11<=22) ]: should succeed.
-iteAtStringOneArg :: Term TyName Name DefaultUni DefaultFun ()
-iteAtStringOneArg = Apply () iteAtString lteExpr
-
 -- [ { (builtin ifThenElse) (forall a . a -> a) } (11<=22) ]: evaluation should
 -- succeed, typechecking should fail (illegal kind)
-iteAtHigherKindOneArg :: Term TyName Name DefaultUni DefaultFun ()
-iteAtHigherKindOneArg = Apply () iteAtHigherKind lteExpr
-
--- (builtin ifThenElse) (11<=22) "11 <= 22" "¬(11<=22)": should succeed.
-iteUninstantiatedFullyApplied :: Term TyName Name DefaultUni DefaultFun ()
-iteUninstantiatedFullyApplied = mkIterApp () ite [lteExpr, stringResultTrue, stringResultFalse]
-
--- [ { (builtin ifThenElse)  (con string) } (11<=22) "11 <= 22" "¬(11<=22)" ]: should succeed.
-iteAtStringFullyApplied :: Term TyName Name DefaultUni DefaultFun ()
-iteAtStringFullyApplied = mkIterApp () iteAtStringOneArg [stringResultTrue, stringResultFalse]
-
--- [ { (builtin ifThenElse) (con integer) } (11<=22) "11 <= 22" "¬(11<=22)" ]:
--- should also succeed.  However it's ill-typed because we're instantiating at
--- `integer` but returning a string: at execution time we only check that type
--- instantiations and term arguments are correctly interleaved, not that
--- instantiations are correct.
-iteAtIntegerFullyApplied :: Term TyName Name DefaultUni DefaultFun ()
-iteAtIntegerFullyApplied = mkIterApp () iteAtIntegerOneArg [stringResultTrue, stringResultFalse]
+iteAtHigherKindWithCond :: Term TyName Name DefaultUni DefaultFun ()
+iteAtHigherKindWithCond = Apply () iteAtHigherKind lteExpr
 
 -- [ {(builtin ifThenElse) (forall a . a -> a) } (11<=22) "11 <= 22" "¬(11<=22) ]":
--- illegal kind, but should succeed.
+-- illegal kind, but should run
```

Comment:
Why is the kind illegal?

Reply:
Because the comment's wrong.  It should say `lam a . a -> a`.  It seems I can't stop myself from writing the wrong thing.